src/main/scala/org/broadinstitute/dsde/firecloud/core/ProfileClient.scala

```diff
@@ -84,77 +82,101 @@ class ProfileClientActor(requestContext: RequestContext) extends Actor with Fire
-
-                            val descSince = DateUtils.prettySince(lastLinkSeconds)
-
-                            val statusResponse = NIHStatus(
-                              loginRequired,
-                              linkedNihUsername = profile.linkedNihUsername,
-                              isDbgapAuthorized = profile.isDbgapAuthorized,
-                              lastLinkTime = Some(lastLinkSeconds),
-                              linkExpireTime = profile.linkExpireTime,
-                              secondsSinceLastLink = Some(secsSinceLastLink),
-                              descriptionSinceLastLink = Some(descSince))
-
-                            RequestComplete(StatusCodes.OK, statusResponse)
-                          case _ =>
-                            // user is linked and authorized, but we have no record of login time or expiration time.
-                            RequestComplete(StatusCodes.OK, NIHStatus(true,
-                              linkedNihUsername = profile.linkedNihUsername,
-                              lastLinkTime = profile.lastLinkTime,
-                              linkExpireTime = profile.linkExpireTime,
-                              isDbgapAuthorized = profile.isDbgapAuthorized))
-
-                        }
-                      case _ => RequestComplete(NoContent, "Not dbGaP authorized")
+    case GetAndUpdateNIHStatus(userInfo: UserInfo) =>
+      getProfile(userInfo, true)
+      // do NOT pipe the response back to the parent!
+  }
+
+  def getProfile(userInfo: UserInfo, updateExpiration: Boolean): Future[PerRequestMessage] = {
+    val pipeline = addCredentials(userInfo.accessToken) ~> sendReceive
+    val profileReq = Get(UserService.remoteGetAllURL.format(userInfo.getUniqueId))
+
+    pipeline(profileReq) map { response: HttpResponse =>
+      response.status match {
+        case OK =>
+          val profileEither = response.entity.as[ProfileWrapper]
+          profileEither match {
+            case Right(profileWrapper) =>
+              val profile = Profile(profileWrapper)
+              profile.linkedNihUsername match {
+                case Some(nihUsername) =>
+                  // we have a linked profile.
+
+                  // if we have no record of the user logging in or the user's expire time, default to 0
+                  val lastLinkSeconds = profile.lastLinkTime.getOrElse(0L)
+                  var linkExpireSeconds = profile.linkExpireTime.getOrElse(0L)
+
+                  val howOld = DateUtils.hoursSince(lastLinkSeconds)
+                  val howSoonExpire = DateUtils.secondsSince(linkExpireSeconds)
+
+                  // if the user's link has expired, the user must re-link.
+                  // NB: we use a separate val here in case we need to change the logic. For instance, we could
+                  // change the logic later to be "link has expired OR user hasn't logged in within 24 hours"
+                  val loginRequired = (howSoonExpire >= 0)
```

Comment:
see the previous code's TODO: we no longer need to check last login time, and we can rely solely on expiration time.